src/analyzer/protocol/vxlan/VXLAN.cc

```diff
@@ -0,0 +1,100 @@
+
+#include "VXLAN.h"
+#include "TunnelEncapsulation.h"
+#include "Conn.h"
+#include "IP.h"
+#include "Reporter.h"
+
+#include "events.bif.h"
+
+using namespace analyzer::vxlan;
+
+void VXLAN_Analyzer::Done()
+	{
+	Analyzer::Done();
+	Event(udp_session_done);
+	}
+
+void VXLAN_Analyzer::DeliverPacket(int len, const u_char* data, bool orig,
+                                   uint64 seq, const IP_Hdr* ip, int caplen)
+	{
+	Analyzer::DeliverPacket(len, data, orig, seq, ip, caplen);
+
+	// Outer Ethernet, IP, and UDP layers already skipped.
+	// Also, generic UDP analyzer already checked/guarantees caplen >= len.
+
+	constexpr auto vxlan_len = 8;
+
+	if ( len < vxlan_len )
+		{
+		ProtocolViolation("VXLAN header truncation", (const char*) data, len);
+		return;
+		}
+
+	if ( (data[0] & 0x08) == 0 )
+		{
+		ProtocolViolation("VXLAN 'I' flag not set", (const char*) data, len);
+		return;
+		}
+
+	const EncapsulationStack* estack = Conn()->GetEncapsulation();
+
+	if ( estack && estack->Depth() >= BifConst::Tunnel::max_depth )
+		{
+		reporter->Weird(Conn(), "tunnel_depth");
+		return;
+		}
+
+	int vni = (data[4] << 16) + (data[5] << 8) + (data[6] << 0);
+
+	data += vxlan_len;
+	caplen -= vxlan_len;
+	len -= vxlan_len;
+
+	pkt_timeval ts;
+	ts.tv_sec = (time_t) current_timestamp;
+	ts.tv_usec = (suseconds_t) ((current_timestamp - (double)ts.tv_sec) * 1000000);
+	Packet pkt(DLT_EN10MB, &ts, caplen, len, data);
+
+	if ( ! pkt.Layer2Valid() )
+		{
+		ProtocolViolation("VXLAN invalid inner ethernet frame",
+		                  (const char*) data, len);
+		return;
+		}
+
+	data += pkt.hdr_size;
+	len -= pkt.hdr_size;
+	caplen -= pkt.hdr_size;
+
+	IP_Hdr* inner = nullptr;
+	int res = 0;
+
+	switch ( pkt.l3_proto ) {
```

Comment:
Hmm, could we send this into `NetSessions::DoNextPacket` here? That would avoid hard-coding another l3_proto switch. Not sure if that works, just wondering.

Reply:
Except I think we do want control back here for the purpose of additional/custom error checking/handling.  E.g. don't want a generic Weird, but rather ProtocolViolation if there's a problem parsing that layer.  Also do want to possibly do more stuff in the case where that layer looks good.  E.g. ProtocolConfirmation and raise further vxlan-specific events.

We might probably still do all that if we further change `NetSessions::DoNextPacket` to allow for it: add return value, suppress/customize weird messages, etc., but seems too early or out of scope.  Other tunnel analysis also already goes through `DoNextInnerPacket`, so it's at least consistent.

Reply:
Ok, sounds like opportunity for a bit of refactoring in the future, possibly along with other L2 work.